x-pack/plugins/security_solution/server/lib/siem_migrations/rules/task/util/rule_resource_retriever.ts

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { isEmpty } from 'lodash/fp';
+import type { QueryResourceIdentifier } from '../../../../../../common/siem_migrations/rules/resources/types';
+import { getRuleResourceIdentifier } from '../../../../../../common/siem_migrations/rules/resources';
+import type {
+  OriginalRule,
+  RuleMigrationResource,
+  RuleMigrationResourceType,
+} from '../../../../../../common/siem_migrations/model/rule_migration.gen';
+import type { RuleMigrationsDataClient } from '../../data/rule_migrations_data_client';
+
+export type RuleMigrationResources = Partial<
+  Record<RuleMigrationResourceType, RuleMigrationResource[]>
+>;
+
+export const MAX_RECURSION_DEPTH = 10;
+
+export class RuleResourceRetriever {
+  constructor(
+    private readonly migrationId: string,
+    private readonly dataClient: RuleMigrationsDataClient
+  ) {}
+
+  public async getResources(originalRule: OriginalRule): Promise<RuleMigrationResources> {
+    const resourceIdentifier = getRuleResourceIdentifier(originalRule);
+    return this.recursiveRetriever(originalRule.query, resourceIdentifier);
+  }
+
+  private recursiveRetriever = async (
+    query: string,
+    resourceIdentifier: QueryResourceIdentifier,
+    it = 0
+  ): Promise<RuleMigrationResources> => {
+    if (it >= MAX_RECURSION_DEPTH) {
+      return {};
```

Comment:
Do we surface some kind of error if we reach this limit to the user? i.e. `we were unable to retrieve all resources or something...`?

Reply:
There should not be more than 2-3 nested levels of resources. This limit is just to prevent infinite recursion in case something goes wrong. I'll add a comment.